.github/workflows/base-ci.yml

```diff
@@ -0,0 +1,115 @@
+name: Shared CI Workflow
+
+on:
+  workflow_call:
+    inputs:
+      use_canary:
+        required: true
```

Comment:
I'm just curious if we can make it optional and canary-ci.yml to call ci.yml. (then, we can avoid base-ci.yml. not a big deal though.)

Reply:
The three file setup is needed to have separated `on:`'s

we could join them into a single config and start to manage conditions with env vars but this feels hacky

something like: 

```yml
name: CI

on:
  push:
    branches: [main]
  pull_request:
    types: [opened, synchronize]
  issue_comment:
    types: [created]

concurrency:
  group: ${{ github.workflow }}-${{ github.ref }}
  cancel-in-progress: true

env:
  IS_CANARY: ${{ github.event.issue.pull_request && startsWith(github.event.comment.body, '/canary-ci run') }}
  SHOULD_RUN: ${{ (github.event_name != 'issue_comment') || (github.event.issue.pull_request && startsWith(github.event.comment.body, '/canary-ci run')) }}

jobs:
  test:
    name: Test on (Node ${{ matrix.version }})
    if: ${{ env.SHOULD_RUN == 'true' }}
    strategy:
      fail-fast: false
      matrix:
        version: [18.17.0, 20.8.0, 22.7.0]
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: pnpm/action-setup@v4
      - uses: actions/setup-node@v3
        with:
          node-version: ${{ matrix.version }}
          cache: 'pnpm'
          cache-dependency-path: '**/pnpm-lock.yaml'
      - name: Install dependencies
        run: |
          ${{ env.IS_CANARY == 'true' && 'pnpm -r update react@canary' || '' }}
          pnpm install --frozen-lockfile
      - run: pnpm test

  build:
    if: ${{ env.SHOULD_RUN == 'true' }}
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: pnpm/action-setup@v4
      - uses: actions/setup-node@v3
        with:
          node-version: 18
          cache: 'pnpm'
          cache-dependency-path: '**/pnpm-lock.yaml'
      - name: Install dependencies
        run: |
          ${{ env.IS_CANARY == 'true' && 'pnpm -r update react@canary' || '' }}
          pnpm install --frozen-lockfile
      - run: pnpm run compile
      - uses: actions/upload-artifact@v4
        with:
          name: create-waku
          path: |
            packages/create-waku
            !packages/create-waku/node_modules
          if-no-files-found: error
      - uses: actions/upload-artifact@v4
        with:
          name: waku
          path: |
            packages/waku
            !packages/waku/node_modules
          if-no-files-found: error

  e2e:
    name: E2E on ${{ matrix.os }} (Node ${{ matrix.version }}) - (${{ matrix.shared }}/4)
    if: ${{ env.SHOULD_RUN == 'true' }}
    needs:
      - build
    strategy:
      fail-fast: false
      matrix:
        shared: [1, 2, 3, 4]
        shardTotal: [4]
        os: [ubuntu-latest, windows-latest, macos-latest]
        version: [18.17.0, 20.8.0, 22.7.0]
        exclude:
          - os: macos-latest
            version: 18.17.0
          - os: macos-latest
            version: 20.8.0
          - os: windows-latest
            version: 18.17.0
          - os: windows-latest
            version: 20.8.0
    runs-on: ${{ matrix.os }}
    steps:
      - uses: actions/checkout@v4
      - uses: pnpm/action-setup@v4
      - uses: actions/setup-node@v3
        with:
          node-version: ${{ matrix.version }}
          cache: 'pnpm'
          cache-dependency-path: '**/pnpm-lock.yaml'
      - name: Install dependencies
        run: |
          ${{ env.IS_CANARY == 'true' && 'pnpm -r update react@canary' || '' }}
          pnpm install --frozen-lockfile
      - uses: actions/download-artifact@v4
        with:
          name: create-waku
          path: packages/create-waku
      - uses: actions/download-artifact@v4
        with:
          name: waku
          path: packages/waku
      - name: Install Playwright
        id: install-playwright
        uses: ./.github/actions/playwright
      - run: pnpm exec playwright test --shard=${{ matrix.shared }}/${{ matrix.shardTotal }}
        env:
          TEMP_DIR: ${{ runner.temp }}
      - uses: actions/upload-artifact@v3
        if: always()
        with:
          name: playwright-report-${{ matrix.os }}-${{ matrix.version }}-${{ matrix.shared }}
          path: test-results/
          retention-days: 30
          if-no-files-found: ignore
```

Reply:
alright. we don't need such a hack for this.